LBH.AdultSocialCare.Functions.Payruns/Services/InvoiceGenerator.cs

```diff
@@ -0,0 +1,121 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using LBH.AdultSocialCare.Functions.Payruns.Enums;
+using LBH.AdultSocialCare.Functions.Payruns.Gateways.Interfaces;
+using LBH.AdultSocialCare.Functions.Payruns.Infrastructure.Entities.CarePackages;
+using LBH.AdultSocialCare.Functions.Payruns.Infrastructure.Entities.Payments;
+using LBH.AdultSocialCare.Functions.Payruns.Services.InvoiceItemGenerators;
+
+namespace LBH.AdultSocialCare.Functions.Payruns.Services
+{
+    public class InvoiceGenerator
+    {
+        private readonly ICarePackageGateway _carePackageGateway;
+        private readonly IFundedNursingCareGateway _fundedNursingCareGateway;
+
+        private Dictionary<PackageType, List<BaseInvoiceItemsGenerator>> _generators;
+
+        public InvoiceGenerator(
+            ICarePackageGateway carePackageGateway, IFundedNursingCareGateway fundedNursingCareGateway)
+        {
+            _carePackageGateway = carePackageGateway;
+            _fundedNursingCareGateway = fundedNursingCareGateway;
+        }
+
+        public async Task<IList<Invoice>> GenerateAsync(IList<CarePackage> packages, DateTimeOffset invoiceEndDate)
+        {
+            await InitializeGeneratorsAsync();
+
+            var invoices = new List<Invoice>();
+            var invoicesCount = await _carePackageGateway.GetInvoicesCountAsync(); // TODO: VK: Review invoice numbering
+
+            foreach (var package in packages)
+            {
+                var coreCost = package.Details.First(d => d.Type is PackageDetailType.CoreCost);
+
+                // TODO: VK: Review and remove
+                var invoiceStartDate = /* package.PaidUpTo ?? */ coreCost.StartDate;
```

Comment:
This will always have core cost start date as invoice start date. No? Last pay run end date would be a little better I think or now the corecost start date if no previous pay run. More important if the date difference here is the one used to get quantity

Reply:
it's come from previous implementation and not elaborated yet, dates handling is the next step

Reply:
Clear